Support query pass-through for Elasticsearch connector

kasiafi opened the pull request:
This PR introduces `raw_query` Polymorphic Table Function for full query pass-through to Elasticsearch connector.

```markdown
# Elasticsearch
* Add `raw_query` table function for full query pass-through to the connector.
```


## Documentation

Documentation issue is filed: https://github.com/trinodb/trino/issues/13007.
When documenting, also deprecate the legacy query pass-through mechanism: https://trino.io/docs/current/connector/elasticsearch.html?highlight=elasticsearch#pass-through-queries

ebyhr commented on plugin/trino-elasticsearch/src/test/java/io/trino/plugin/elasticsearch/BaseElasticsearchConnectorTest.java:
```diff
@@ -1860,6 +1860,12 @@ public void testMissingIndex()
+    @Test
+    public void testRemoteQueryTableFunction()
+    {
+        assertQuerySucceeds("SELECT * FROM TABLE(\"" + getSession().getCatalog().orElseThrow() + "\".\"system\".\"remote_query\"(\"schema\" => 'tpch', \"index\" => 'nation', \"query\" => '{\"query\": {\"match\": {\"name\": \"ALGERIA\"}}}'))");
```
It would be better to verify the result in my opinion and how about formatting like this? 
```java
        assertThat(
                "SELECT * FROM " +
                        "TABLE(\"" + getSession().getCatalog().orElseThrow() + "\".\"system\".\"remote_query\"(" +
                        " \"schema\" => 'tpch'," +
                        " \"index\" => 'nation'," +
                        " \"query\" => '{\"query\": {\"match\": {\"name\": \"ALGERIA\"}}}'" +
                        "))")
                .matches(...);
```

Also, I would recommend adding empt result & failure case.

martint replied:
That's intentional. Pass-through queries in Elasticsearch return a single VARCHAR column as a result.

kasiafi replied:
@ebyhr I added more tests including no match and failure.

martint commented on plugin/trino-elasticsearch/src/main/java/io/trino/plugin/elasticsearch/ptf/RemoteQuery.java:
```diff
@@ -0,0 +1,132 @@
+public class RemoteQuery
+        implements Provider<ConnectorTableFunction>
+{
+    public static final String NAME = "remote_query";
```
Call this `query`

martint commented on plugin/trino-elasticsearch/src/main/java/io/trino/plugin/elasticsearch/ElasticsearchMetadata.java:
```diff
@@ -563,8 +568,8 @@ public Optional<ConstraintApplicationResult<ConnectorTableHandle>> applyFilter(C
                         if (metadata.getSchema()
-                                    .getFields().stream()
-                                    .anyMatch(field -> columnName.equals(field.getName()) && field.getType() instanceof PrimitiveType && "keyword".equals(((PrimitiveType) field.getType()).getName()))) {
+                                .getFields().stream()
+                                .anyMatch(field -> columnName.equals(field.getName()) && field.getType() instanceof PrimitiveType && "keyword".equals(((PrimitiveType) field.getType()).getName()))) {
```
Unrelated formatting change?

martint reviewed (approved):
We should deprecate the old-style passthrough queries in favor of this.

martint commented on plugin/trino-elasticsearch/src/main/java/io/trino/plugin/elasticsearch/ptf/Query.java:
```diff
@@ -0,0 +1,144 @@
+        }
+    }
+
+    public static class QueryHandle
```
`QueryFunctionHandle` maybe?

martint commented on plugin/trino-elasticsearch/src/test/java/io/trino/plugin/elasticsearch/BaseElasticsearchConnectorTest.java:
```diff
@@ -1860,6 +1861,69 @@ public void testMissingIndex()
+                .hasMessageContaining("json_parse_exception");
+    }
+
+    private void assertQueryMatchesAny(String actual, List<String> expectedOptions)
+    {
+        for (String expected : expectedOptions) {
+            try {
+                assertQuery(actual, expected);
+                return;
+            }
+            catch (AssertionError e) {
+                // ignore, try other options
+            }
+        }
+        throw new AssertionError("Neither of expected options matched the actual result");
+    }
```
I don't think this is such a good idea. It masks the underlying assertion, so it's hard to tell what was different and it what way. It also doesn't scale when there are more than a couple of alternatives.

Instead, modify the query to normalize the results in some way. For instance, parse the resulting json, extract the values and sort them to get a consistent result.

martint reviewed (commented):
Also, don't forget to update the docs or file an issue to do it later so we don't lose track of it.

kasiafi commented:
@martint I applied comments. Please take a look.

findepi commented on plugin/trino-elasticsearch/src/test/java/io/trino/plugin/elasticsearch/BaseElasticsearchConnectorTest.java:
```diff
@@ -1860,6 +1861,53 @@ public void testMissingIndex()
+    {
+        // select single record
+        assertQuery("SELECT json_query(result, 'lax $[0][0].hits.hits._source') " +
+                        "FROM TABLE(elasticsearch.system.query(" +
+                        "schema => 'tpch', " +
+                        "index => 'nation', " +
+                        "query => '{\"query\": {\"match\": {\"name\": \"ALGERIA\"}}}')) t(result)",
+                "VALUES '{\"nationkey\":0,\"name\":\"ALGERIA\",\"regionkey\":0,\"comment\":\" haggle. carefully final deposits detect slyly agai\"}'");
```
due to the fact that the function seems to dump ES result, i would consider calling a `raw_query`
(especially so that we keep `query` name free for future)

martint replied:
That's a good point. In ES parlance, queries are "search queries" (https://www.elastic.co/guide/en/elasticsearch/reference/current/search-your-data.html), and there's a Query DSL (https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl.html) to formulate them (what we support here). They also have two other languages:
* EQL, for querying log and event data (https://www.elastic.co/guide/en/elasticsearch/reference/current/eql.html)
* SQL (https://www.elastic.co/guide/en/elasticsearch/reference/current/sql-overview.html)

kasiafi commented:
@martint @findepi I renamed the function to `raw_query`.

I also added a test  to compare the results between the legacy pass-through mechanism, and the new table function.

The old query pass-through will be deprecated in the documentation when we document the table function. @martint is there any other way we want to deprecate it than a note in the docs?
